File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/system/OnBootReceiver.java
package nerd.tuxmobil.fahrplan.congress.system;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import java.util.List;

import info.metadude.android.eventfahrplan.commons.logging.Logging;
import info.metadude.android.eventfahrplan.commons.temporal.Moment;
import nerd.tuxmobil.fahrplan.congress.alarms.AlarmReceiver;
import nerd.tuxmobil.fahrplan.congress.alarms.AlarmServices;
import nerd.tuxmobil.fahrplan.congress.autoupdate.UpdateService;
import nerd.tuxmobil.fahrplan.congress.dataconverters.AlarmExtensions;
import nerd.tuxmobil.fahrplan.congress.models.Alarm;
import nerd.tuxmobil.fahrplan.congress.models.SchedulableAlarm;
import nerd.tuxmobil.fahrplan.congress.repositories.AppRepository;
import nerd.tuxmobil.fahrplan.congress.utils.FahrplanMisc;

public final class OnBootReceiver extends BroadcastReceiver {

    private static final String LOG_TAG = "OnBootReceiver";
    private final Logging logging = Logging.get();

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent == null) {
            return;
        }
        String action = intent.getAction();
        if (action == null) {
            return;
        }
        if (!action.equals(AlarmReceiver.ALARM_UPDATE)) {
            return;
        }

        logging.report(LOG_TAG, "onReceive (reboot)");

        Moment nowMoment = Moment.now().plusSeconds(15);

        AppRepository appRepository = AppRepository.INSTANCE;
        List<Alarm> alarms = appRepository.readAlarms();
        AlarmServices alarmServices = AlarmServices.newInstance(context, appRepository);
        for (Alarm alarm : alarms) {
            Moment storedAlarmTime = Moment.ofEpochMilli(alarm.getStartTime());
            if (nowMoment.isBefore(storedAlarmTime)) {
                logging.d(LOG_TAG, "Scheduling alarm for session: " + alarm.getSessionId() + ", " + alarm.getSessionTitle());
                SchedulableAlarm schedulableAlarm = AlarmExtensions.toSchedulableAlarm(alarm);
                alarmServices.scheduleSessionAlarm(schedulableAlarm);
            } else {
                logging.d(LOG_TAG, "Deleting alarm from database: " + alarm);
                appRepository.deleteAlarmForAlarmId(alarm.getId());
            }
        }

        // start auto updates
        boolean isAutoUpdateEnabled = appRepository.readAutoUpdateEnabled();
        if (isAutoUpdateEnabled) {
            long lastFetchedAt = appRepository.readScheduleLastFetchedAt();
            long nowMillis = Moment.now().toMilliseconds();

            long interval = FahrplanMisc.setUpdateAlarm(context, true, logging);

            logging.d(LOG_TAG, "now: " + nowMillis + ", lastFetchedAt: " + lastFetchedAt);
            if (interval > 0 && nowMillis - lastFetchedAt >= interval) {
                UpdateService.start(context);
            }
        }
    }

}

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/system/OnBootReceiver.kt
package nerd.tuxmobil.fahrplan.congress.system

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import info.metadude.android.eventfahrplan.commons.logging.Logging
import info.metadude.android.eventfahrplan.commons.temporal.Moment
import nerd.tuxmobil.fahrplan.congress.alarms.AlarmServices
import nerd.tuxmobil.fahrplan.congress.autoupdate.UpdateService
import nerd.tuxmobil.fahrplan.congress.dataconverters.toSchedulableAlarm
import nerd.tuxmobil.fahrplan.congress.repositories.AppRepository
import nerd.tuxmobil.fahrplan.congress.utils.FahrplanMisc

class OnBootReceiver : BroadcastReceiver() {

    private companion object {
        const val LOG_TAG = "OnBootReceiver"
    }

    private val logging = Logging.get()

    override fun onReceive(context: Context, intent: Intent) {
        val action = intent.action ?: return
        if (action != Intent.ACTION_BOOT_COMPLETED) {
            return
        }

        logging.report(LOG_TAG, "onReceive (reboot)")
        val nowMoment = Moment.now().plusSeconds(15)
        val appRepository = AppRepository.apply {
            initialize(
                context = context.applicationContext,
                logging = logging,
            )
        }
        val alarmServices = AlarmServices.newInstance(context, appRepository)
        val alarms = appRepository.readAlarms()
        for (alarm in alarms) {
            // Check if the alarm time has passed
            if (nowMoment.isBefore(Moment.ofEpochMilli(alarm.startTime))) {
                logging.d(LOG_TAG, "Scheduling alarm for session: ${alarm.sessionTitle}, ${alarm.sessionTitle}")
                alarmServices.scheduleSessionAlarm(alarm.toSchedulableAlarm())
            } else {
                logging.d(LOG_TAG, "Deleting alarm from database: $alarm")
                appRepository.deleteAlarmForAlarmId(alarm.id)
            }
        }

        // Start auto updates
        if (appRepository.readAutoUpdateEnabled()) { // Check if auto update is enabled
            val lastFetchedAt = appRepository.readScheduleLastFetchedAt()
            val nowMillis = Moment.now().toMilliseconds()
            val interval = FahrplanMisc.setUpdateAlarm(
                context = context,
                conferenceTimeFrame = appRepository.loadConferenceTimeFrame(),
                isInitial = true,
                logging = logging,
                onCancelScheduleNextFetch = appRepository::deleteScheduleNextFetch,
                onUpdateScheduleNextFetch = appRepository::updateScheduleNextFetch,
            )

            logging.d(LOG_TAG, "now: $nowMillis, lastFetchedAt: $lastFetchedAt")

            if (interval > 0 && nowMillis - lastFetchedAt >= interval) {
                UpdateService.start(context)
            }
        }
    }

}
